lib/charms/mongodb/v1/shards_interface.py

```diff
@@ -691,8 +711,36 @@ def _on_relation_changed(self, event):
 
         self.charm.app_peer_data["mongos_hosts"] = json.dumps(self.get_mongos_hosts())
 
-    def pass_hook_checks(self, event):
+    def pass_hook_checks(self, event: EventBase):
         """Runs the pre-hooks checks for ConfigServerRequirer, returns True if all pass."""
+        proceed_complex_hook_checks = self.pass_sanity_hook_checks(event)
+        if not proceed_complex_hook_checks:
+            return False
+
+        # occasionally, broken events have no application, in these scenarios nothing should be
+        # processed.
+        if not event.relation.app and isinstance(event, RelationBrokenEvent):
+            return False
```

Comment:
TBH it has been a long time since I have written this code and I am not too familiar with this use case, but I recall this check being necessary.

Thanks a ton for this context and info, it is super helpful

This code was just moved around in this PR. Perhaps we could discuss this at a later date to dig into why this issue was occurring, but not in this PR since it's related to sharding and not upgrades.